docs: rework the "Events" section of world api.md (#5012)
ArchipelagoMW/Archipelago: docs/world api.md

Reviewer: I read through this and it makes more sense to me than the original.  Having multiple examples is very helpful.  After reading though it did leave me with one gap, which is what situations you might use an event where you didn't place_locked_item and what that would look like.

Author: At the moment I think getting into even more complex use cases would be going too far beyond the scope of this document. I could see something like this in apworld_dev_faq, though I'm not sure this is "frequent" enough.

Though to directly answer the question: You could have a set of event items and event locations and decide which items go on which locations based on some player options or even random choice. This would presumably only make sense if you were also modifying the game (using slot_data or a patch file) to make sure that reshuffling of whatever in-game actions the events represent matches the generation-time shuffling. Personally, I suspect most situations in which you could modify the game enough to make that work are also situations where you'd rather make regular locations instead of events, but it's theoretically viable.

An item rule is a function that returns `True` or `False` for a `Location` based on a single item. It can be used to
reject the placement of an item there.

### Events (or "generation-only items/locations")

An event item or location is one that only exists during multiworld generation. The server is never made aware of them,
event locations can never be checked, nor can event items be received during play. Events are used to represent in-game

> actions (that aren't regular Archipelago locations) when either:

Author: Weird, I thought I checked the rest of this file and found item/location were lowercase _except_ in the old Events section, so I'd been making them lowercase here for consistency.

> actions (that aren't regular Archipelago locations) when either:

Reviewer 2: Oh yeah I think you're right, my bad

> actions (that aren't regular Archipelago locations) when either:

Author: I should probably also mention that the only reason I was thinking about this in the first place is that the old section said "Event" with a capital E more than once, and I realized that's confusing because there is no `Event` class (unlike all the other words for core AP concepts that we sometimes capitalize).

Then I found myself looking at "event Item" and "event Location" in my early drafts, which of course forced me to ask if Item/Location really needed to be capitalized, so here we are.

* we want to show in the spoiler log when the player is expected to perform the in-game action, or
* it's the cleanest way to represent how that in-game action impacts logic

Typical examples include completing the goal, defeating a boss, or flipping a switch that changes several areas.

To be precise: the term "event" on its own refers to the special combination of an "event item" placed on an "event
location". Event items and locations are created the same way as normal items and locations, except that they have an

> `id` of `None`, they must not exist in the `name_to_id` lookups, and an event item must be placed on an event location

Reviewer 2: Is this actually true, regarding them not being allowed to be in the name_to_id lookups? I was under the impression that you could just set the ID of an item/location to None to effectively turn it into an event.

Author: Good question. I was trusting that the existing docs are correct about this.

Author: If I actually put `{ "Event": None }` in my `item_name_to_id` map, it gets quietly deleted before my world methods get called. So no problems arise in practice, but I don't know if this "supported" behavior or "core working around a broken world" behavior.

Reviewer 2: Does it break if you change a Location (one you physically create) to have an ID of None, without touching location_name_to_id? and I guess also locking some item onto it while also giving it an ID of None without touching item_name_to_id. I'd check but I'm in the office still

Author: If I did this correctly, the answer is that doing that breaks the tests for my item/location name groups, `test_default_all_state_can_reach_everything` fails claiming the location is "unreachable", `test_two_player_single_game_fills` fails for the same reason, and most of my world tests also fail for the same reason.

Reviewer 3: > If I actually put `{ "Event": None }` in my `item_name_to_id` map, it gets quietly deleted before my world methods get called. So no problems arise in practice, but I don't know if this "supported" behavior or "core working around a broken world" behavior.

it's "core working around a broken world behavior but when it was proposed to change to not do this workaround it was worse code overall so it still does the workaround even if it's against spec"

Reviewer 3: but yes you can have a Location/Item that has an id of None and has a name in the datapackage, just don't put those name: None mappings in your datapackage

Reviewer 2: > but yes you can have a Location/Item that has an id of None and has a name in the datapackage, just don't put those name: None mappings in your datapackage

Okay great, that's what I thought was the case

> (and vice versa). By convention, the event location and item typically have the same name, though this is not a
> requirement. Finally, although events are often described as "fake" items and locations, it's important to understand

Reviewer 2: ```suggestion
(and vice versa). Finally, although events are often described as "fake" items and locations, it's important to understand
```
I don't think this is a good convention -- since it gets shown in the spoiler log, there should be an attempt to make it more readable than just "Victory: Victory"

Author: Honestly, I agree. But I felt trying to change recommendations/policy would be out of scope in a PR about making these docs easier to understand.

Reviewer 2: I'm one of the docs maintainers so you have my approval to make this change o:

Author: All right then. I've pushed another commit with just the "different names" changes.

Author: This must be some kind of record; it still hasn't processed the commit

![image](https://github.com/user-attachments/assets/932df79a-0364-4e29-90a8-92fd8b468c07)

that they are perfectly real during generation.

The most common way to create an event is to create the event item and the event location, then immediately call
`.place_locked_item()`:

```python
victory_loc = MyGameLocation(self.player, "Victory", None)
victory_loc.place_locked_item(MyGameItem("Victory", ItemClassification.progression, None, self.player))
self.multiworld.completion_condition[self.player] = lambda state: state.has("Victory", self.player)
set_rule(victory_loc, lambda state: state.can_reach_region("Final Boss Arena", self.player))
```

Having a `"Victory"` event item that's required to finish the game makes it very clear in the spoiler log when the

> player finishes, rather than only seeing their last relevant item. But events aren't just about the spoiler log;

Reviewer 2: ```suggestion
player is expected to finish, rather than only seeing their last relevant item. But events aren't just about the spoiler log;
```

a more substantial example of using events to structure your logic might be:

```python
water_loc = MyGameLocation(self.player, "Lowered Water Level", None)
water_loc.place_locked_item(MyGameItem("Lowered Water Level", ItemClassification.progression, None, self.player))
set_rule(water_loc, lambda state: state.can_reach_region("Water Pump Switch", self.player))
...
basement_loc = MyGameLocation(self.player, "Flooded House - Basement Chest", None)
flooded_house_region.add_locations([upstairs_loc, ground_floor_loc, basement_loc], MyGameLocation)
...
set_rule(basement_loc, lambda state: state.has("Lowered Water Level", self.player))
```

> This creates a "Lowered Water Level" event with an access rule, and a regular location whose access rule depends on that

Reviewer 2: ```suggestion
This creates a "Lowered Water Level" event and a regular location whose access rule depends on that
```

An item rule is a function that returns `True` or `False` for a `Location` based on a single item. It can be used to
reject the placement of an item there.

### Events (or "generation-only items/locations")

An event item or location is one that only exists during multiworld generation. The server is never made aware of them,
event locations can never be checked, nor can event items be received during play.

Events are used to represent in-game actions (that aren't regular Archipelago locations) when either:

* We want to show in the spoiler log when the player is expected to perform the in-game action.
* It's the cleanest way to represent how that in-game action impacts logic.

Typical examples include completing the goal, defeating a boss, or flipping a switch that affects multiple areas.

To be precise: the term "event" on its own refers to the special combination of an "event item" placed on an "event
location". Event items and locations are created the same way as normal items and locations, except that they have an

> `id` of `None`, they must not exist in the `name_to_id` lookups, and an event item must be placed on an event location

Author: Huh, I drew the exact opposite conclusion from that discussion. I could see softening the wording to "should not", but breaking several unit tests and being "against spec" seem like enough evidence to keep this clause?

> `id` of `None`, they must not exist in the `name_to_id` lookups, and an event item must be placed on an event location

Author: Talked on Discord, and we realized we were reading this phrase in two very different ways (having `None`s directly in the `name_to_id` maps, versus contradicting the maps by generating a "real" location with an id of `None`). So now I'm in favor of deleting the phrase just because it's ambiguous, and neither meaning is important enough to discuss explicitly.

* It's the cleanest way to represent how that in-game action impacts logic.

Typical examples include completing the goal, defeating a boss, or flipping a switch that affects multiple areas.

To be precise: the term "event" on its own refers to the special combination of an "event item" placed on an "event
location". Event items and locations are created the same way as normal items and locations, except that they have an
`id` of `None` and an event item must be placed on an event location
(and vice versa). Finally, although events are often described as "fake" items and locations, it's important to
understand that they are perfectly real during generation.

The most common way to create an event is to create the event item and the event location, then immediately call
`.place_locked_item()`:

```python
victory_loc = MyGameLocation(self.player, "Defeat the Final Boss", None, final_boss_arena_region)
victory_loc.place_locked_item(MyGameItem("Victory", ItemClassification.progression, None, self.player))
self.multiworld.completion_condition[self.player] = lambda state: state.has("Victory", self.player)

> set_rule(victory_loc, lambda state: state.can_reach_region("Final Boss Arena", self.player))

Reviewer 2: ```suggestion
set_rule(victory_loc, lambda state: state.has("Boss Defeating Sword", self.player))
```
`state.can_reach_region` on a location's parent region will never return false, and we definitely don't want the docs to have an example like that.

```

Requiring an event to finish the game will make the spoiler log display an additional
`Defeat the Final Boss: Victory` line when the player is expected to finish, rather than only showing their last
relevant item. But events aren't just about the spoiler log; a more substantial example of using events to structure
your logic might be:

```python

> water_loc = MyGameLocation(self.player, "Water Level Switch", None)

Reviewer 3: these should also have the parent region in the constructor otherwise location.parent_region is unset (missed this in my first review) ((applies to all of the codeblocks))

> water_loc = MyGameLocation(self.player, "Water Level Switch", None)
> water_loc.place_locked_item(MyGameItem("Lowered Water Level", ItemClassification.progression, None, self.player))
> pump_station_region.add_locations([water_loc], MyGameLocation)
> set_rule(water_loc, lambda state: state.has("Double Jump", self.player))  # the switch is really high up
> ...
> basement_loc = MyGameLocation(self.player, "Flooded House - Basement Chest", None)
> flooded_house_region.add_locations([upstairs_loc, ground_floor_loc, basement_loc], MyGameLocation)

Reviewer 3: ```suggestion
water_loc = MyGameLocation(self.player, "Water Level Switch", None)
water_loc.place_locked_item(MyGameItem("Lowered Water Level", ItemClassification.progression, None, self.player))
pump_station_region.locations.append(water_loc)
set_rule(water_loc, lambda state: state.has("Double Jump", self.player))  # the switch is really high up
...
basement_loc = MyGameLocation(self.player, "Flooded House - Basement Chest", None)
flooded_house_region.locations += [upstairs_loc, ground_floor_loc, basement_loc]
```
region.add_locations is for when you don't instantiate the objects yourselves (which is why it takes names/ids and a Location Class)

  If it's hard to separate, this can be done during `generate_early` or `create_items` as well.
* `create_items(self)`
  called to place player's items into the MultiWorld's itempool. By the end of this step all regions, locations and
  items have to be in the MultiWorld's regions and itempool (including events). You cannot add or remove items,

>   locations, or regions after this step. Locations cannot be moved to different regions after this step.

Author: Good point. Though this phrasing might imply events are something separate from items and locations. I'll try something else.
